src/audio/google_rtc_audio_processing.c

```diff
@@ -0,0 +1,387 @@
+
+	return comp_set_state(dev, cmd);
+}
+
+static int google_rtc_audio_processing_prepare(struct comp_dev *dev)
+{
+	struct google_rtc_audio_processing_comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *source_buffer;
+	struct list_item *source_buffer_list_item;
+	uint32_t flags;
+
+	comp_dbg(dev, "google_rtc_audio_processing_prepare()");
+
+	/* searching for stream and feedback source buffers */
+	list_for_item(source_buffer_list_item, &dev->bsource_list) {
+		source_buffer = container_of(source_buffer_list_item, struct comp_buffer,
+						 sink_list);
+		buffer_lock(source_buffer, &flags);
+		if (source_buffer->source->comp.type == SOF_COMP_DEMUX)
+			cd->aec_reference = source_buffer;
+		else if (source_buffer->source->comp.type == SOF_COMP_DAI)
+			cd->raw_microphone = source_buffer;
+		buffer_unlock(source_buffer, flags);
+	}
+
+	cd->output = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+
+	switch (cd->output->stream.frame_fmt) {
+#if CONFIG_FORMAT_S16LE
+	case SOF_IPC_FRAME_S16_LE:
+		break;
+#endif /* CONFIG_FORMAT_S16LE */
+	default:
+		comp_err(dev, "unsupported data format: %d", cd->output->stream.frame_fmt);
+		return -EINVAL;
+	}
+
+	if (cd->output->stream.rate != GOOGLE_RTC_AUDIO_PROCESSING_SAMPLERATE) {
+		comp_err(dev, "unsupported samplerate: %d", cd->output->stream.rate);
+		return -EINVAL;
+	}
+
+	return comp_set_state(dev, COMP_TRIGGER_PREPARE);
+}
+
+static int google_rtc_audio_processing_reset(struct comp_dev *dev)
+{
+	comp_dbg(dev, "google_rtc_audio_processing_reset()");
+
+	return comp_set_state(dev, COMP_TRIGGER_RESET);
+}
+
+static int google_rtc_audio_processing_copy(struct comp_dev *dev)
+{
+	struct google_rtc_audio_processing_comp_data *cd = comp_get_drvdata(dev);
+
+	struct comp_copy_limits cl;
+	int frame;
+	uint32_t aec_reference_buff_frag = 0;
+		uint32_t raw_micriphone_buff_frag = 0;
+		uint32_t output_buff_frag = 0;
+	int16_t *src;
+	int16_t *dst;
+	uint32_t flags = 0;
+
+	buffer_lock(cd->aec_reference, &flags);
+	const uint32_t num_aec_reference_frames = audio_stream_get_avail_frames(&cd->aec_reference->stream);
+	const uint32_t num_aec_reference_bytes = audio_stream_get_avail_bytes(&cd->aec_reference->stream);
+	buffer_unlock(cd->aec_reference, flags);
+
+	buffer_invalidate(cd->aec_reference, num_aec_reference_bytes);
+
+	aec_reference_buff_frag = 0;
+	for (frame = 0; frame < num_aec_reference_frames; frame++) {
+		src = audio_stream_read_frag_s16(&cd->aec_reference->stream, aec_reference_buff_frag);
+		cd->aec_reference_buffer[cd->aec_reference_buffer_index] = *src;
+		++cd->aec_reference_buffer_index;
+
+		if (cd->aec_reference_buffer_index == cd->num_frames) {
+			GoogleRtcAudioProcessingAnalyzeRender_int16(cd->state, cd->aec_reference_buffer);
+			cd->aec_reference_buffer_index = 0;
+		}
+		aec_reference_buff_frag += cd->aec_reference->stream.channels;
+	}
+	comp_update_buffer_consume(cd->aec_reference, num_aec_reference_bytes);
+
+
```

Comment:
don't think double empty lines are used a lot in SOF. Personally I prefer to avoid them to maximise the monitor real estate use. One empty line usually provides enough separation